### DISCUSSION TITLE
Sub Invoice Lines in Mustang

### DISCUSSION THREAD

**Discussion by fktrdui** [ZUGFeRD/mustangproject#381] [Q&A] [answered]
Hello,
I'm currently exploring the capabilities of the Mustang Project, particularly in the context of xrechnung invoicing. 
As I dive deeper into the documentation and codebase, I encountered a query regarding the representation of XRechnung Extensions and sub invoice lines. In many invoicing scenarios, it's common to have nested or hierarchical structures within an invoice, where one line item might contain further breakdowns or details.
Could you kindly provide insights into how the Mustang Project handles such scenarios? Are there specific classes, objects, or attributes designated for representing sub invoice lines? 
Thank you for your time and assistance.

**Comment by jstaerk** [comment, marked as answer]
Hi,
Mustang's was created to support Factur-X, i.e. Cross Industry Invoice (CII).
 
Within CII Sub invoice lines would be possible using either ParentLineID (to an infinte depth) or IncludedSubordinateTradeLineItem (to a maximum depth of 1). ParentLineID would be almost trivial to implement and not require any changes in any validator. However, as far as I know Kosit so far never publish a mapping to CII for their extension profile. 

Once they specify this one could implement it and support their "extension" profile just like any Factur-X profile. But for the time being we do not yet support UBL output (#320), no Factur-X profile supports either ParentLineID nor IncludedSubordinateTradeLineItem, and Kosit supports sub invoice lines only in UBL, so currently there is no support in Mustang.

kind regards
Jochen

**Reply by fktrdui** [reply]
Hi Jochen,
thank you for providing clarity on the current status of sub invoice line support within the Mustang Project.

If I understand correctly, the 'referencedLineID' field in the 'Item' Class corresponds to the 'ParentLineID' concept. However, upon inspecting the 'LineCalculator' class, I couldn't find any calculation functions specifically addressing referenced items. Should this be handled differently, or is there another method or class where this calculation is managed?

Regarding IncludedSubordinateTradeLineItem, I'm eager to learn more about its implementation and how I can overwrite its value. Could you please provide additional information on this aspect?

Best Regards,
Farzam

**Reply by jstaerk** [reply]
Hello,
no, the concept is not yet implemented, the referencedLineID refers to line IDs in orders the invoice line refers to.

Calculatory-wise nothing would have to be changed, however, because, unlike the IncludedSubordinateTradeLineItem concept parentLineID does not provide redundant values, i.e. if you have a 200000€ house you would do it like this and give the parent line a value of 0 (or any other value, which adds to the invoice total). In the XML you'd have
``` 
ID parent  name         amount
1   n/a       house        0
2   1          roof        100000
3   1          basement    100000
``` 
Any viewer could do the math itself and the viewing software could show

``` 
+ house 200000

``` 
and if you open it it would show 100000€ roof and basement. I'm not totally sure how the IncludedSubordinateTradeLineItem (or any UBL) concept is but this has the advantage that you can decide if your parent line *has* a value, and which one, and sub lines are independendent in terms of they could even have different tax rates and/or charges / allowances. 

I'd love to provide further info, unfortunately I hardly have any, apart from the official schema and schematron files and https://unece.org/fileadmin/DAM/uncefact/BRS/BRS_CII_v2.0.6.pdf . However, in Mustang it is not yet supported because neither Factur-X nor XRechnung/CII refer to this concept yet, and as long as the Xrechnung does not explain how to represent it in CII and I see no use of supporting a CII concept which fails on *any* validator simply because the attributes/elements (e.g. parentLineID), may they be present in the CII schema files, they are not present in Factur-X schema files or schematron checks.

Kind regards
Jochen
